<launch>
    <!-- LAUNCH INTERFACE -->
    <arg name="use_gazebo_gui" default="true" />
    <arg name="robot_name" default="vito"/>
    <arg name="use_rviz" default="false"/>
    <arg name="use_joint_state_publisher" default="true"/>
    <arg name="use_grasp_estimator" default="false"/>

    <arg name="left_arm_enabled" default="false"/>
    <arg name="right_arm_enabled" default="false"/>

    <!-- choose whether or not to start the full robot simulation -->
    <arg name="use_robot_sim" default="true"/>
    <arg name="publish_frequency" default="500"/>
    <param name="publish_frequency" value="$(arg publish_frequency)"/>

    <!-- LAUNCH IMPLEMENTATION -->

    <!-- #### -->

    <!-- load the full robot in robot_description, useful for visualization, collision detection and planning simultaneously -->
    <!-- the selection is only due to the hand visualization which is different in real and simulation -->
    <group if="$(arg use_robot_sim)">
        <param name="robot_description" command="$(find xacro)/xacro --inorder $(find dual_manipulation_control)/robot/$(arg robot_name).urdf.xacro use_mimic_tag:=true"/>
    </group>
    <group unless="$(arg use_robot_sim)">
        <param name="robot_description" command="$(find xacro)/xacro --inorder $(find dual_manipulation_control)/robot/$(arg robot_name).urdf.xacro use_mimic_tag:=true"/>
    </group>

    <group if="$(arg use_joint_state_publisher)">
        <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
            <param name="use_gui" value="false"/>
            <param name="rate" value="$(arg publish_frequency)"/>
            <rosparam param="source_list">[/left_arm/joint_states,/left_hand/joint_states,/right_arm/joint_states,/right_hand/joint_states,/head/joint_states]</rosparam>
        </node>
    </group>
    
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"  />

    <group if="$(arg use_rviz)">
        <node name="my_rviz" pkg="rviz" type="rviz" respawn="false" args="-d $(find dual_manipulation_control)/launch/rviz_config.rviz" output="screen"/>
    </group>

    <rosparam command="load" file="$(find dual_manipulation_control)/config/vito_controllers.yaml"/>
    <rosparam command="load" file="$(find dual_manipulation_control)/config/hardware_interfaces.yaml"/>


    <group if="$(arg use_robot_sim)">
        <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param robot_description -urdf -model $(arg robot_name) -x 0 -y 0 -z 1.0 -R 0 -P 0 -Y 0" respawn="false" output="screen"/>

        <include file="$(find gazebo_ros)/launch/empty_world.launch">
            <arg name="world_name" value="$(find dual_manipulation_control)/worlds/simple_environment.world"/>
            <arg name="paused" value="false"/>
            <arg name="use_sim_time" value="true"/>
            <arg name="gui" value="$(arg use_gazebo_gui)"/>
            <arg name="headless" value="true"/>
            <arg name="debug" value="false"/>
        </include>

        <param name="cube_description" command="$(find xacro)/xacro --inorder $(find dual_manipulation_control)/urdf/cube.urdf.xacro" />
        <node name="cube_spawner" pkg="gazebo_ros" type="spawn_model" args="-param cube_description -urdf -model cube -x -0.8 -y 0 -z 2.15 -R 0 -P 0 -Y 0" respawn="false" output="screen"/>
    </group>

    <group ns="left_arm">
        <group if="$(arg left_arm_enabled)">
            <param name="robot_description" command="$(find xacro)/xacro --inorder $(find dual_manipulation_control)/robot/$(arg robot_name).urdf.xacro use_mimic_tag:=true"/>

            <rosparam command="load" file="$(find dual_manipulation_control)/config/left_arm_names.yaml" />

            <include file="$(find lwr_hw)/launch/lwr_hw.launch">
                <arg name="port" value="49948"/>
                <arg name="ip" value="192.168.0.20"/>
                <arg name="name" value="left_arm"/>
            </include>
        </group>
        <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen"  args="joint_state_controller joint_impedance_controller"/>
    </group>
    

    <group ns="right_arm">
        <group if="$(arg right_arm_enabled)">
            <param name="robot_description" command="$(find xacro)/xacro --inorder $(find dual_manipulation_control)/robot/$(arg robot_name).urdf.xacro use_mimic_tag:=true"/>
            <rosparam command="load" file="$(find dual_manipulation_control)/config/right_arm_names.yaml" />

            <include file="$(find lwr_hw)/launch/lwr_hw.launch">
                <arg name="port" value="49939"/>
                <arg name="ip" value="192.168.0.10"/>
                <arg name="name" value="right_arm"/>
            </include>
        </group>

        <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen"  args="joint_state_controller joint_impedance_controller"/>
    </group>

    <!-- set gains for the controller-->
    <param name="/kp" value="-500.0"/>
    <param name="/kv" value="-1.0"/>

</launch>
